Sandrine's weather app

MindstormingAB opened the pull request:
Link to my deployed project: https://smhi-killer.netlify.app/

My reflections on how this week's project turned out: It has been really exciting to work with APIs! I am still struggling a bit with objects and arrays but I usually manage to solve my issues after some time reflection (and lots of console logs!)

Things I'd like to have clarified or explained in more detail: How to hide the API key

IngelaL commented on code/style.css:
```diff
@@ -0,0 +1,162 @@
+    flex-direction: column;
+    justify-content: center;
+    margin: 0 5%;
+    width: 35%;
```
Great job with the media queries. Your weather app looks great and have 3 different styles depending on device. 💯

IngelaL commented on code/style.css:
```diff
@@ -0,0 +1,162 @@
+
+@media (min-width: 768px) {
+  body {
+  display: flex;
+  flex-direction: column;
+  justify-content: center;
+  align-items: center;
```
Add indentation here.

MindstormingAB replied:
Thanks! 💐

IngelaL commented:
Sandrine, reviewing your code have been so inspiring. You really know what you are doing and your project is at an extremely high level.  To me this is flawless and you should be so proud of yourself. The best advice I can give you is to add even more comments in your code so that you can help others to follow and learn from it, and for you to go back to. Great job! 💪

MindstormingAB commented on code/script.js:
```diff
@@ -0,0 +1,247 @@
+  'Thu',
+  'Fri',
+  'Sat'
+]
```
Useful feedback, I will complete my code with some comments 👍

MindstormingAB commented on code/script.js:
```diff
@@ -0,0 +1,247 @@
+  const timeZone = forecastArray.city.timezone;
+  const filteredForecast = forecastArray.list.filter(item => item.dt_txt.includes('12:00'));
+  const forecastWeather = filteredForecast.map(day => {
+    const dayName = (new Date((day.dt + timeZone - 7200) * 1000)).toLocaleDateString([], {
```
I think that the issue is caused by this `[]` on line 190; I omitted to set a language format. It doesn't matter on my computer that is in English but I guess yours is in Swedish 😕. I'm going to improve my code thanks to your comment. 💐

MindstormingAB commented on code/script.js:
```diff
@@ -0,0 +1,247 @@
+// This function starts page loading with current location
+getLocation();
+
+// This function displays today's weather
+const displayTodaysWeather = (weatherArray) => {
+  city.innerHTML = weatherArray.name;
+  const timeStamp = new Date((weatherArray.dt + weatherArray.timezone - 7200) * 1000);
+  time.innerHTML = `${days[timeStamp.getDay()]} ${timeStamp.getDate()} ${months[timeStamp.getMonth()]} ${timeStamp.toLocaleTimeString([], {
+    hour: '2-digit',
+    minute: '2-digit',
+    hour12: false,
+  })}`;
+  const temp = weatherArray.main.temp;
+  temperature.innerHTML = `${temp.toFixed(1)}\xB0`;
+  const descriptionArray = weatherArray.weather.map((a) => a.main);
+  description.innerHTML = descriptionArray[0];
+  const iconArray = weatherArray.weather.map((a) => a.icon);
+  const todayIcon = iconArray[0];
+  returnBodyBackground(todayIcon);
+  returnTempGradient(temp);
+  picto.src = 'https://openweathermap.org/img/wn/' + todayIcon + '@2x.png';
+  sunrise.innerHTML = (new Date((weatherArray.sys.sunrise + weatherArray.timezone - 7200) * 1000)).toLocaleTimeString([], {
+    hour: '2-digit',
+    minute: '2-digit',
+    hour12: false,
+  });
+  sunset.innerHTML = (new Date((weatherArray.sys.sunset + weatherArray.timezone - 7200) * 1000)).toLocaleTimeString([], {
+    hour: '2-digit',
+    minute: '2-digit',
+    hour12: false,
+  });
```
I've added comments as separators

MindstormingAB commented on code/script.js:
```diff
@@ -0,0 +1,247 @@
+const returnWeatherByCity = (cityValue) => {
+  const WEATHER_API_URL = `https://api.openweathermap.org/data/2.5/weather?q=${cityValue}&units=metric&APPID=` + WEATHER_API_KEY;
+  const FORECAST_API_URL = `https://api.openweathermap.org/data/2.5/forecast?q=${cityValue}&units=metric&APPID=` + WEATHER_API_KEY;
+  fetch(WEATHER_API_URL)
```
Very good feedback. I started writing the second function by copying the first one and didn't notice the names were identical! 💐

MindstormingAB commented on code/index.html:
```diff
@@ -0,0 +1,84 @@
+
+  </main>
+
+  <script defer src="./script.js"></script>
```
I stole this attribute from Van's code during lecture. According to https://www.w3schools.com/tags/att_script_defer.asp, it means that the script is executed when the page has finished parsing. Not completely sure why it is good 😬

MindstormingAB commented:
@IngelaL  Thank you so muck for your feedback. I have updated my code accordingly.

EmOlsz commented on code/index.html:
```diff
@@ -0,0 +1,84 @@
+<body>
+
+  <header>
+    <section class="citySelector">
```
classes should me written in `kebab-case`, not `camelCase` - needs to be fixed in whole project

EmOlsz commented on code/index.html:
```diff
@@ -0,0 +1,84 @@
+
+    <section class="citySearch">
+      <input id="input" type="text" placeholder="Type a city name">
+      <button id="city-search">Search</button>
```
I think you might mistake id's with classes when it comes to naming convention :)

- All classes should written in `kebab-case`
- All id's can be written in `camelCase`

MindstormingAB replied:
Thank you Maks! I got a bit nervous when I first read your comment: I have a certain number of class names and ids! 😖 But luckily it was only in the header I dropped the naming convention! 😌

EmOlsz commented on code/script.js:
```diff
@@ -0,0 +1,261 @@
+]
+
+// This function returns body background image depending on weather icon code
+const returnBodyBackground = (todayIcon) => {
```
If we pass only one parameter inside, we can omit parentheses 

```
const returnBodyBackground = todayIcon => { ... }
```

EmOlsz commented on code/script.js:
```diff
@@ -0,0 +1,261 @@
+
+// This function defines current location and passes information to showPosition()
+const getLocation = () => {
+  navigator.geolocation.getCurrentPosition(showPosition);
```
Great job :)

EmOlsz commented on code/script.js:
```diff
@@ -0,0 +1,261 @@
+}
+
+// This function updates weather information based on city name selected by user
+selectButton.addEventListener('click', () => {
```
👍

EmOlsz commented on code/style.css:
```diff
@@ -0,0 +1,162 @@
+}
+
+header {
+  background: rgba(142,158,171,1);
```
You don't need to use `rgba()` if you set last argument (opacity) to 1 - it's set as that in shorter version, `rgb()`

MindstormingAB replied:
@EmOlsz Is your comment even valid when it comes to gradient backgrounds? See line 13 for instance: can I change `rgba` to `rgb` and remove opacity or is `rgba` mandatory in gradients?

EmOlsz commented on code/style.css:
```diff
@@ -0,0 +1,162 @@
+    box-shadow:10px 10px 5px 0px rgba(20,20,20,0.75);
+    top: auto;
+    right: auto;
+    left: auto;
```
Those 3 lines of code don't do anything at all - try to write code that you fully understand, otherwise it can lead you to many unpredicted situations :)

MindstormingAB replied:
You got me there! I struggled a lot with the positioning of body, main and header in relation to each others and how to make it consistent regardless screen size. I experimented with these 3 properties and forgot to remove them 🤯

EmOlsz replied:
;)

EmOlsz reviewed (changes requested):
Well done project Sandrine, a lot of extra requirements done :) Great job, truly!
You didn't avoid mistakes though, please take a look on my comments and provide changes. I hope they will help you for the future!

MindstormingAB commented:
Hi Maks! Thank you so much for your encouraging comments. 
I fixed the names of the classes and ids where I switched kebab-case and camelCase, removed some parentheses in one parameter functions, changed `rgba` to `rgb` when `rgba` was not needed and removed unused code.
